third_party/move/move-vm/runtime/src/move_vm.rs

```diff
@@ -3,22 +3,98 @@
-}
+/// Move VM is completely stateless. It is used to execute a single loaded function with its type
+/// arguments fully instantiated.
+pub struct MoveVm;
 
-impl MoveVM {
-    /// Creates a new VM instance.
-    #[allow(clippy::new_without_default)]
-    pub fn new() -> Self {
-        Self {
-            runtime: VMRuntime::new(),
-        }
+impl MoveVm {
+    pub(crate) fn execute_loaded_function(
+        function: LoadedFunction,
+        serialized_args: Vec<impl Borrow<[u8]>>,
+        data_store: &mut TransactionDataCache,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        extensions: &mut NativeContextExtensions,
+        module_storage: &impl ModuleStorage,
+    ) -> VMResult<SerializedReturnValues> {
+        let vm_config = module_storage.runtime_environment().vm_config();
+        let ty_builder = &vm_config.ty_builder;
+
+        let create_ty_with_subst = |tys: &[Type]| -> VMResult<Vec<Type>> {
+            tys.iter()
+                .map(|ty| ty_builder.create_ty_with_subst(ty, function.ty_args()))
+                .collect::<PartialVMResult<Vec<_>>>()
+                .map_err(|err| err.finish(Location::Undefined))
+        };
+
+        let param_tys = create_ty_with_subst(function.param_tys())?;
+        let (mut dummy_locals, deserialized_args) =
+            deserialize_args(module_storage, &param_tys, serialized_args)
+                .map_err(|e| e.finish(Location::Undefined))?;
+
+        let return_tys = create_ty_with_subst(function.return_tys())?;
+
+        let timer = VM_TIMER.timer_with_label("Interpreter::entrypoint");
+        let return_values = Interpreter::entrypoint(
+            function,
+            deserialized_args,
+            data_store,
+            module_storage,
+            gas_meter,
+            traversal_context,
+            extensions,
+        )?;
+        drop(timer);
+
+        let return_values = serialize_return_values(module_storage, &return_tys, return_values)
+            .map_err(|e| e.finish(Location::Undefined))?;
+        let mutable_reference_outputs = param_tys
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, ty)| match ty {
+                Type::MutableReference(inner_ty) => Some((idx, inner_ty.as_ref())),
+                _ => None,
+            })
+            .map(|(idx, ty)| {
+                // serialize return values first in the case that a value points into this local
+                let local_val =
+                    dummy_locals.move_loc(idx, vm_config.check_invariant_in_swap_loc)?;
+                let (bytes, layout) = serialize_return_value(module_storage, ty, local_val)?;
+                Ok((idx as LocalIndex, bytes, layout))
+            })
+            .collect::<PartialVMResult<_>>()
+            .map_err(|e| e.finish(Location::Undefined))?;
+
+        // locals should not be dropped until all return values are serialized
+        drop(dummy_locals);
```

Comment:
any benefit to explicitly dropping here?

Also, it's a lot cleaner now - great refactor. it's moved so a bit harder to review side by side (nothing two windows can't fix). But please do point out when things aren't just copy-paste (e.g. add a comment inline) so reviewers don't miss.